kernel: invoke init:run_on_load_handlers/0 after code_server start

vngthb opened the pull request:
Moving the invocation of init:on_load_handlers/0 from kernel:init(safe) to code:do_start/0.

init:on_load_handlers/0 has been added as part of https://github.com/erlang/otp/commit/7f82edd and it was implemented in such a way that it will be invoked, quoting, "after having started most servers in the kernel application", problem is that some of those servers, under specific configurations (see #6085), need the on load handlers in order for them to properly start.

As mentioned, this should fix #6085.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/erlang/otp?pullRequest=6227) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/erlang/otp?pullRequest=6227) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/erlang/otp?pullRequest=6227) it.</sub>

bjorng reviewed (changes requested):
Thanks for your pull request.

Please sign the [Contributor License Agreement](https://cla-assistant.io/erlang/otp?pullRequest=6227).

With this pull request, the test case [code_SUITE:bit_boot_embedded/1](https://github.com/erlang/otp/blob/d079ee1e20cb289fa858849cd4cfdf2f689dee91/lib/kernel/test/code_SUITE.erl#L1198-L1206) fails. The reason is that the on_load handler for the `dyntrace` module expects the file server process to be started.

I think that a better place to run the on_load handlers would be in `kernel_config:start_link/0`. Can you do that change and verify that it solves #6085?

When you have done that, please update the commit message to contain the same information as the pull request description. We want anyone that want to know about the change in 10 years time to find the information directly in the commit message instead of having to locate the pull request.

bjorng commented:
@rickard-green Could there be a problem for on_load handlers if the `kernel_refc` process is not started? If so, could the `kernel_refc` server be started before the `kernel_config process`?

@garazdawi The logger supervisor is started before the `kernel_config` process in `minimal` mode, but after the `safe` supervisor in other modes. I suppose there is a good reason for that? Can there be any consequences if an on_load handler is called before the logger supervisor is started?

vngthb commented:
> Please sign the [Contributor License Agreement](https://cla-assistant.io/erlang/otp?pullRequest=6227).

Done!

> With this pull request, the test case [code_SUITE:bit_boot_embedded/1](https://github.com/erlang/otp/blob/d079ee1e20cb289fa858849cd4cfdf2f689dee91/lib/kernel/test/code_SUITE.erl#L1198-L1206) fails. The reason is that the on_load handler for the `dyntrace` module expects the file server process to be started.
> 
> I think that a better place to run the on_load handlers would be in `kernel_config:start_link/0`. Can you do that change and verify that it solves #6085?

Made the suggested change and it does solve it, updated the commit.

> When you have done that, please update the commit message to contain the same information as the pull request description. We want anyone that want to know about the change in 10 years time to find the information directly in the commit message instead of having to locate the pull request.

Updated the commit message as well.

bjorng commented:
Thanks! Added to our daily builds.

garazdawi commented:
> @garazdawi The logger supervisor is started before the `kernel_config` process in `minimal` mode, but after the `safe` supervisor in other modes. I suppose there is a good reason for that? Can there be any consequences if an on_load handler is called before the logger supervisor is started?

I think it should be fine. Any log messages sent during on_load will be handled by the simple logger and get logged correctly when the logger supervisor is started.

garazdawi commented:
Hmm, one of the problems with this change is that any on_load callback will no longer have access to standard_out or standard_error. So any call to `io:format()` will crash. Logging will still work as it falls back to erlang:display if needed...

I'm unsure how common it is print to stdout/stderr when on_load is called. All OTP code uses logger to report any errors.

garazdawi commented:
I've also found that #6144 needs this PR as the tty nif has to be loaded before user is started in the kernel supervision tree.

garazdawi commented:
I looked a bit closer at this, and at least for the new tty nif added in #6144 I've decided that the best way is to extend init to allow calling of on_load for a specific module earlier in the boot sequence.

You can view the changes I did [here](https://github.com/erlang/otp/pull/6144/commits/eb0165e169a628f19e92a7ab9fc060db6e18db5d). If we are unsure if moving the place where on_load are called then we could make the tls dist module call the same init function for crypto.

garazdawi commented:
Me and @bjorng have talked a bit and agreed that the solution is to use the [init function](https://github.com/erlang/otp/pull/6144/commits/cf981d3849a8b737fc2a7be44bc86835015ced5d) I've introduced in #6144 for loading crypto if needed for the distribution.

I will push new commits to this PR when I've merged that PR so that you can test that they work in your system before merging.

garazdawi commented:
I've updated your branch with the changes needed and added a test case to make sure that it works. Please try it and make sure that it also solves your problem.

@IngelaAndin @RaimoNiskanen I would appreciate any comments you have on the contents of this PR.

vngthb commented:
@garazdawi, sorry for the late response, it does indeed solve the issue described in #6085, thank you!

garazdawi commented:
Thanks for the trouble report and the PR!
